BUG: fix SLSQP max iteration setting to fix #4921

AtsushiSakai opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
To fix #4921 

#### What does this implement/fix?
fix SLSQP max iteration setting and enable SLSQP and TNC unittest.

#### Additional information
TNC unittest can pass the unittest in current master branch code, so I enabled all methods for the test.

andyfaff commented:
Id like to think about this for a few days.

andyfaff commented:
Thinking a bit more deeply I think there are a few things that need to be changed in the SLSQP, and one discussion point in TNC.

**SLSQP**
Could you check when maxiter is reached that `OptimizeResult.status == 9`. Also the status message needs to be changed on L308, from `Iteration Limit Exceeded` to `Iteration Limit Reached`. Also the docstring for `fmin_slsqp` needs to be changed. I'm not sure if there are other locations in the documentation where those exit mode messages are written.

**TNC**
In the context of tnc the `maxiter` parameter actually refers to the maximum number of function evaluations (L355 of tnc.py). I don't think it's actually possible at the moment to set a max number of iterations, just a max number of function evaluations.

andyfaff commented on scipy/optimize/slsqp.py:
```diff
@@ -242,7 +242,7 @@ def _minimize_slsqp(func, x0, args=(), jac=None, bounds=None,
     """
     _check_unknown_options(unknown_options)
-    iter = maxiter
+    iter = maxiter-1
```
PEP8: space between `maxiter` and 1.

AtsushiSakai replied:
Thanks. I fixed it.

AtsushiSakai commented:
Hi. @andyfaff . Thank you for your review.  I added a test for OptimizeResult check and updated exit code message and docstring. Please take a look.

I have a question on TNC. 
Number of iterations and number of function evaluations are not same meaning in TNC optimization?

andyfaff commented:
When I inspected the TNC code it looks like `maxiter` is taken to mean the maximum number of iterations. From the docstring:
```
    maxiter : int
        Maximum number of function evaluation. If None, `maxiter` is
        set to max(100, 10*len(x0)). Defaults to None.
```
and:
```
In[8]: minimize(rosen, [2,3,2], method='tnc', options={'maxiter': 10})
Out[8]: 
     fun: 191.67628671396187
     jac: array([ 119.82580832, -532.93195793,  271.2364124 ])
 message: 'Max. number of function evaluations reached'
    nfev: 44
     nit: 2
```

This is confusing, and I think it's a bug. Normally setting maximum function iterations would use a `maxfun` kwd. Note that I think `res.nit`, and `res.nfev` are correct.

andyfaff commented on scipy/optimize/slsqp.py:
```diff
@@ -172,7 +172,7 @@ def fmin_slsqp(func, x0, eqcons=(), f_eqcons=None, ieqcons=(), f_ieqcons=None,
          7 : Rank-deficient equality constraint subproblem HFTI
          8 : Positive directional derivative for linesearch
-         9 : Iteration limit exceeded
+         9 : Iteration limit Reached
```
reached does not require capitalisation.

AtsushiSakai replied:
I fixed it.

andyfaff replied:
`Reached` --> `reached`

AtsushiSakai replied:
Oh. sorry. I fixed another one. I fixed this one.

andyfaff commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -937,6 +936,10 @@ def test_respect_maxiter(self, method):
         if hasattr(sol, 'njev'):
             assert sol.njev >= sf.ngev
 
+        # method specific tests
```
I'm wondering if we should set line 934 to `sol.nit == MAXITER`. I suspect tnc would then fail tests.

AtsushiSakai replied:
Oh. you are right.  The test failed. sol.nit is 1, not 4.

> ============================================================================================= FAILURES ==============================================================================================
> ___________________________________________________________________________ TestOptimizeSimple.test_respect_maxiter[TNC] ____________________________________________________________________________
> scipy/optimize/tests/test_optimize.py:935: in test_respect_maxiter
>     assert sol.nit == MAXITER
> E   assert 1 == 4
> E    +  where 1 =      fun: 8.799939396486105\n     jac: array([-1.9414383 , -1.34732119,  0.37293528,  0.38459742, -0.09909762,\n        ...08, 0.0121846 , 0.0121846 , 0.0121846 , 0.00979179,\n       0.0121846 , 0.0121846 , 0.0121846 , 0.0145774 , 0.02632083]).nit

AtsushiSakai replied:
I changed "maxiter" to "maxfun" in options of tnc solver.

andyfaff commented:
> I changed "maxiter" to "maxfun" in options of tnc solver.

I'd like to get feedback from other maintainers here, because it might cause problems for a few of our users.

@mdhaber

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -932,11 +935,15 @@ def test_respect_maxiter(self, method):
             assert sol.njev >= sf.ngev
 
+        # method specific tests
+        if method in ('SLSQP',):
```
I assume you're setting up to add more of these. Although this collection has only one element, a set seems more appropriate to look for something in.
```suggestion
        if method in {'SLSQP'}:
```
Otherwise, why not `if method == 'SLSQP'`?

AtsushiSakai replied:
if method == 'SLSQP' is fine. I fixed it.

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1092,8 +1099,12 @@ def hess(x):
 
             for f, g, h in itertools.product(funcs, grads, hesss):
                 count = [0]
+                if method in ['tnc']:
```
```suggestion
                if method in {'tnc'}:
```

AtsushiSakai replied:
i fixed it.

andyfaff commented:
@mdhaber my ping was due to using maxfun in TNC rather than maxiter.

mdhaber commented:
I haven't read TNC too closely, but from the discussion above it sounds like

-  the maximum number of function evaluations and the maximum number of iterations are distinct for TNC, 
- that the current interface allows control over the maximum number of _function evaluations_ using the `maxiter` option, and 
- that there is no way to actually control the number of iterations.

I agree that

- if `maxiter` actually sets the max number of function evaluations and 
- if the number of function evaluations does not equal the number of iterations

then `maxfun` should have been the name of the option.

I don't think _just_ changing the name now is a good idea. I think one of the following would be better:

- Leave the name alone
- Add `maxfun`, have it do the same thing as `maxiter`, and mention both in the documentation. 
- (best) Do what you have done, but also add the ability to control the maximum number of iterations with `maxiter`

AtsushiSakai commented:
@mdhaber @andyfaff Thank you for reviewing. I added maxiter also for tnc solver. Currently, I prioritizes maxiter over maxfun. I rolled back some test codes because maxiter can be used. Please take a look.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -284,7 +290,7 @@ def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
 def _minimize_tnc(fun, x0, args=(), jac=None, bounds=None,
                   eps=1e-8, scale=None, offset=None, mesg_num=None,
-                  maxCGit=-1, maxiter=None, eta=-1, stepmx=0, accuracy=0,
+                  maxCGit=-1, maxfun=None, maxiter=None, eta=-1, stepmx=0, accuracy=0,
```
When new keywords are added to a function they should be added at the end of the keyword list.

AtsushiSakai replied:
I moved it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -311,9 +317,12 @@ def _minimize_tnc(fun, x0, args=(), jac=None, bounds=None,
         max(1,min(50,n/2)). Defaults to -1.
-    maxiter : int
-        Maximum number of function evaluation. If None, `maxiter` is
+    maxfun : int
```
move `maxfun` to the end of the docstring, as the keyword is going to the end of the argument list.

AtsushiSakai replied:
I moved it.

andyfaff replied:
You moved maxiter, not maxfun

AtsushiSakai replied:
Oh, sorry. I fixed.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -152,6 +152,9 @@ def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
     maxfun : int, optional
         Maximum number of function evaluation. If None, maxfun is
         set to max(100, 10*len(x0)). Defaults to None.
+    maxiter : int, optional
```
I don't think that we need to add `maxiter` to the keyword arguments.

AtsushiSakai replied:
I removed it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -90,7 +90,7 @@
 def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
              bounds=None, epsilon=1e-8, scale=None, offset=None,
-             messages=MSG_ALL, maxCGit=-1, maxfun=None, eta=-1,
+             messages=MSG_ALL, maxCGit=-1, maxfun=None, maxiter=None, eta=-1,
```
I don't think that we need to add `maxiter` to the keyword arguments.

AtsushiSakai replied:
I removed it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -284,9 +287,10 @@ def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
 def _minimize_tnc(fun, x0, args=(), jac=None, bounds=None,
                   eps=1e-8, scale=None, offset=None, mesg_num=None,
-                  maxCGit=-1, maxiter=None, eta=-1, stepmx=0, accuracy=0,
+                  maxCGit=-1, maxfun=None, eta=-1, stepmx=0, accuracy=0,
```
leave this line as `maxCGit=-1, maxiter=None, eta=-1, stepmx=0, accuracy=0,` (i.e. no change)

AtsushiSakai replied:
I did it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -284,9 +287,10 @@ def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
+                  maxCGit=-1, maxfun=None, eta=-1, stepmx=0, accuracy=0,
                   minfev=0, ftol=-1, xtol=-1, gtol=-1, rescale=-1, disp=False,
-                  callback=None, finite_diff_rel_step=None, **unknown_options):
+                  callback=None, finite_diff_rel_step=None, maxiter=None,
```
`maxiter` --> `maxfun`

AtsushiSakai replied:
I fixed it.

AtsushiSakai commented on scipy/optimize/tnc.py:
```diff
@@ -244,6 +244,9 @@ def fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0,
     SIAM Journal of Numerical Analysis 21, pp. 770-778
 
     """
+    if (maxfun is None) and (maxiter is not None):
```
Oh. you are right. These lines do not needed anymore. I removed it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -311,9 +312,9 @@ def _minimize_tnc(fun, x0, args=(), jac=None, bounds=None,
-        Maximum number of function evaluation. If None, `maxiter` is
-        set to max(100, 10*len(x0)). Defaults to None.
+    maxiter : int, optional
+        Maximum number of function evaluation. If `maxfun` is None, this is
```
Could we change this to:
```
Maximum number of function evaluations. This keyword is deprecated in favor of `maxfun`. Only if `maxfun` is None is this keyword used.

AtsushiSakai replied:
I fixed it.

andyfaff commented on scipy/optimize/tnc.py:
```diff
@@ -350,9 +351,11 @@ def _minimize_tnc(fun, x0, args=(), jac=None, bounds=None,
         the sign of `h` is ignored. If None (default) then step is selected
         automatically.
+    maxfun : int
+        Maximum number of function evaluation. If None, `maxfun` is
```
evaluation -->evaluations.

AtsushiSakai commented:
@andyfaff Should I fix other something?

AtsushiSakai commented:
@andyfaff I did it. PTAL.

andyfaff commented:
Looks good to me now. Thank you @AtsushiSakai for all the hard work.

mdhaber commented:
And thanks @andyfaff for finishing this off; it dropped off my radar. You know, I just discovered the GitHub notification center / inbox today? : )
